### ISSUE TITLE
RAUC firmware_update workflow state was executed twice by thin-edge.io

### ISSUE THREAD

**Issue by reubenmiller** [thin-edge/thin-edge.io#2896]
**Describe the bug**

**Note:** This is part of an investigation and it is not 100% what the exact cause is, or if the bug description is fully accurate.

A log from the RAUC firmware update workflow, shows that after a coming up from a reboot (when switching partitions during the update), that (each?) subsequent state is executed twice (instead of once) by thin-edge.io. See the log in the "additional context" section, but below is a snippet showing the duplicated "firmware_update @ commit" state (which is execute ~2 seconds after another)

```log
----------------------[ firmware_update @ commit | time=2024-05-22T20:33:14.02442458Z ]----------------------

State:    {"logPath":"/data/tedge/log/agent/workflow-firmware_update-c8y-mapper-99685.log","name":"core-image-tedge-rauc","nextPartition":"B","partition":"A","reason":"Install was successful, rebooting device into spare partition","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/98972","status":"commit","tedgeUrl":"http://127.0.0.1:8001/c8y/inventory/binaries/98972","url":"http://127.0.0.1:8001/c8y/inventory/binaries/98972","version":"20240522.2151"}

Action:   /usr/bin/rauc_workflow.sh commit --firmware-name core-image-tedge-rauc --firmware-version 20240522.2151 --url https://t9679.latest.stage.c8y.io/inventory/binaries/98972 --expected-partition B

Exit status: 0 (OK)

stderr <<EOF
2024-05-22T20:33:14 [current=commit] Executing: rauc status mark-good
2024-05-22T20:33:15 [current=commit] Commit successful. New default partition is B
2024-05-22T20:33:16 [current=commit] Saving firmware info to /etc/tedge/.firmware
EOF

stdout <<EOF
rauc status: marked slot rootfs.1 as good
EOF

----------------------[ firmware_update @ commit | time=2024-05-22T20:33:16.415202226Z ]----------------------

State:    {"logPath":"/data/tedge/log/agent/workflow-firmware_update-c8y-mapper-99685.log","name":"core-image-tedge-rauc","nextPartition":"B","partition":"A","reason":"Install was successful, rebooting device into spare partition","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/98972","status":"commit","tedgeUrl":"http://127.0.0.1:8001/c8y/inventory/binaries/98972","url":"http://127.0.0.1:8001/c8y/inventory/binaries/98972","version":"20240522.2151"}

Action:   /usr/bin/rauc_workflow.sh commit --firmware-name core-image-tedge-rauc --firmware-version 20240522.2151 --url https://t9679.latest.stage.c8y.io/inventory/binaries/98972 --expected-partition B

Exit status: 0 (OK)

stderr <<EOF
2024-05-22T20:33:16 [current=commit] Executing: rauc status mark-good
2024-05-22T20:33:17 [current=commit] Commit successful. New default partition is B
2024-05-22T20:33:18 [current=commit] Saving firmware info to /etc/tedge/.firmware
EOF

stdout <<EOF
rauc status: marked slot rootfs.1 as good
EOF

```

It is unclear if this occurs every time, or only sporadically.

Fortunately, the RAUC thin-edge.io workflow does not fail if commit is executed twice (as the commit, in RAUC land, is a "mark-good" command which can be executed multiple times on a partition), however other tooling will fail on this step (e.g. Mender will have a problem as it can only do a "commit" once, and subsequent calls to commit will fail).

**To Reproduce**

The steps to reproduce are not confirmed, but I'll at least describe what was done.

1. Flash a memory card for a Raspberry Pi 4 device using the RAUC enabled Yocto image. The *.wic.bz2 file to flash to the card can be found here under the [core-image-tedge-rauc_raspberrypi4-64" artifact](https://github.com/thin-edge/meta-tedge-project/actions/runs/9196956595)
2. Boot the device
3. Bootstrap the device to get it to connect with Cumulocity IoT

    For example if your device is called "mydevice", then use:

    ```sh
    c8y tedge bootstrap root@mydevice.local
    ```

    You can also scan for it using:

    ```sh
    c8y tedge scan
    ```

4. Upload a [second RAUC firmware update artifact](https://github.com/thin-edge/meta-tedge-project/actions/runs/9202878389) (e.g. the `*.raucb` file) to Cumulocity IoT under the Device Management > Firmware Repository section

    ```sh
    # Create the new name of firmware (just need to do this once)
    c8y firmware create --name "core-image-tedge-rauc"

    # Upload the firmware file (to create a version)
    c8y firmware versions create --firmware core-image-tedge-rauc --file ./core-image-tedge-rauc-raspberrypi4-64_20240523.0212.raucb --version 20240523.0212
    ```

5. Using the Cumulocity IoT device management UI, do a firmware update on the device using the firmware version uploaded in the previous version

    Note: be patient with the firmware update, it can take ~5-10 mins depending on the device's internet connection (and SD card write speed)

6. Inspect the workflow log file afterwards

    In the RAUC yocto image, the thin-edge.io log files are stored under the `/data` partition (for persistence across A/B updates). You should be able to use the following one-liner to print the contents of the most recent firmware_update log file.

    ```sh
    cat "$(ls -t /data/tedge/log/agent/*firmware_update* | head -n1)"
    ```

**Expected behavior**

Each state should only be execute once (unless if the workflow definition is doing some looping, but that is up to the workflow itself)

**Screenshots**

**Environment (please complete the following information):**

|Property|Value|
|--------|-----|
|OS [incl. version]|`Poky (Yocto Project Reference Distro) 4.0.18 (kirkstone)`|
|Hardware [incl. revision]|`Raspberry Pi 4 Model B Rev 1.1`|
|System-Architecture|`Linux rpi4-dca632486720 5.15.92-v8 #1 SMP PREEMPT Wed Feb 8 16:47:50 UTC 2023 aarch64 GNU/Linux`|
|thin-edge.io version|`tedge 1.1.0`|

**Additional context**

The full workflow log can be found here:

[rpi4-d83add90fe56_firwmware_update-c8y-mapper-99704.txt](https://github.com/thin-edge/thin-edge.io/files/15413000/rpi4-d83add90fe56_firwmware_update-c8y-mapper-99704.txt)

**Comment by didier-wenzek** [comment]
Looking at the message sequence gives important insights.

```
$ grep '\[ firmware_update' rpi4-d83add90fe56_firwmware_update-c8y-mapper-99704.txt | cut -d '[' -f 2 | cut -d ']' -f 1 | nl

     1	 firmware_update @ init | time=2024-05-22T20:28:01.146003004Z 
     2	 firmware_update @ scheduled | time=2024-05-22T20:28:01.162339072Z 
     3	 firmware_update @ executing | time=2024-05-22T20:28:01.178696214Z 
     4	 firmware_update @ download | time=2024-05-22T20:28:03.373278528Z 
     5	 firmware_update @ install | time=2024-05-22T20:28:06.617066321Z 
     6	 firmware_update @ restart | time=2024-05-22T20:30:30.754041646Z 
     7	 firmware_update @ await_restart | time=2024-05-22T20:30:30.773915118Z 
     8	 firmware_update > restart @ init | time=2024-05-22T20:30:30.786721218Z 
     9	 firmware_update > restart @ scheduled | time=2024-05-22T20:30:30.804862829Z 
    10	 firmware_update > restart @ executing | time=2024-05-22T20:30:30.831531586Z 
    11	 firmware_update > restart @ executing | time=2022-04-28T17:42:32.332603032Z 
    12	 firmware_update @ await_restart | time=2022-04-28T17:42:32.338678569Z 
    13	 firmware_update > restart @ successful | time=2022-04-28T17:42:32.390875421Z 
    14	 firmware_update @ restarted | time=2022-04-28T17:42:32.404680606Z 
    15	 firmware_update @ restarted | time=2024-05-22T20:31:59.043723562Z 
    16	 firmware_update @ verify | time=2024-05-22T20:32:01.412197265Z 
    17	 firmware_update @ verify | time=2024-05-22T20:33:12.729950117Z 
    18	 firmware_update @ commit | time=2024-05-22T20:33:14.02442458Z 
    19	 firmware_update @ commit | time=2024-05-22T20:33:16.415202226Z 
    20	 firmware_update @ successful | time=2024-05-22T20:33:18.829955029Z 
    21	 firmware_update @ successful | time=2024-05-22T20:33:18.853155636Z 
```

- Event 10 (`restart @ executing`) is published just before the agent triggers a reboot 
- Event 11 (`restart @ executing`) is published just after the agent restarts, reloading from disk the pending commands
  - It might be good to improve the log, to make a difference between the two `restart @ executing` events.
  - First surprise: even 11 is received after event 10 (there is a causal relationship),
    but the timestamp of the second is one month older!
    => the device clock jumped backward on reboot.
- Event 12, is also reloaded from disk by the agent on restart (as all the pending commands).
  - It might be good to improve the log, to stress that this here a no-op:
    the `firmware_update > restart` sub operation is not finalized yet (will be with event 13).
  - It might also be good to __add a new entry in the log__ (between 13 and 14),
    when the state `firmware_update @ await_restart` is considered again,
    but that time to move forward and proceed with `firmware_update @ restarted`.
  - I suspect __the bug is here__, event 12 being actually effective
    the check for the restart being delayed a few tenths of a second,
    just enough to kick off concurrently with the second check (unfortunately not logged).
- What is wrong are events 14 and 15.
  - They are duplicates of a state (`firmware_update @ restarted`) that should be executed only once.
  - The timestamp of the second event jumps one month forward => NTP kicked off.
- Then we have two concurrent threads following the same workflow
  - Since two (`firmware_update @ restarted`) have been generated, this is not surprising.
  - However, __the agent should be more robust__ and stop with an error sooner.

**Comment by reubenmiller** [comment]
> the device clock jumped backward on reboot.

If a device does not have a battery for the clock, then this is expected/normal behaviour.

**Comment by didier-wenzek** [comment]
> If a device does not have a battery for the clock, then this is expected/normal behaviour.

Sure. And this is not a problem: event time is only used for logging and  in no case by the workflow progression logic.

**Comment by gligorisaev** [comment]
Bug tested and is not reproducable anymore.
- added test steps to:
`tests/RobotFramework/tests/tedge_agent/workflows/custom_operation.robot`

The script first extracts relevant log entries from the specified log file, It tries to match the extracted log entries against an expected log file, If the first expected log file does not match, it attempts to match against a second expected log file, this ensures the log file follows one of the two expected sequences.
